### PR TITLE
fix(datastore): Change outbox comparison to allow partial matching

### PR DESCRIPTION
**Pull request by stocaaro**
#### Description of changes
**Brief:** When multiple object updates for a single DataStore object are queued in the outbox to be synced to AppSync, comparison logic is overly strict and as a result only the first update is successfully applied. As a result Datastore appears to accept each update before reverting to the first update made while the outbox was queuing.

We have a number of tests the exercise variations on the problem and may be helpful to review. [Example](https://github.com/aws-amplify/amplify-js/blob/main/packages/datastore/__tests__/connectivityHandling.test.ts#L1129-L1180)

**Context:** When model changes are queued up in the outbox, `_version` field updates are passed on to other updates to the same object that are queued up in the outbox **IF** the object being processed is exactly and deeply equal to the return value from the AppSync update.

**Problem:** Updates send partial models to AppSync, for which AppSync returns complete models. These complete models may contain the same fields that are sent for update, but by also including all other fields, these fail the equality match and as a result we don't update the other objects queued up in the outbox. Since these objects don't receive the updated `_version`, they are rejected by AppSync conflict resolution and the object in Datastore reverts to the last saved content.

**Proposed fix:** Instead of comparing all fields, compare only the fields being updated when checking equality before updating object versions in the outbox.

**Example:**
Object sent to AppSync as an update (when updating only the title
`{title: 'post title 0', id: '1a94e616-506c-47f0-8dbc-db81f264e7e8'}`

Object returned from AppSync after successfully updating
`{id: '1a94e616-506c-47f0-8dbc-db81f264e7e8', title: 'post title 0', order: 348, altText: 'Alt Text - 617', blogPostsId: '52ba510d-5233-4866-a531-1563291d72b1', …}`

While these are obviously not an exact full equality match, all fields from the input object are exactly the same as the output object.

#### Description of how you validated changes
- Updated unit tests
- Added tests around the new `objectMatches` comparison function
- Tested in a sample application

#### Checklist
- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/__tests__/connectivityHandling.test.ts [hunk 1]
```diff
@@ -2187,9 +2186,8 @@ describe('DataStore sync engine', () => {
 			};
 
 			await resyncWith([
-				syncExpression(
-					LegacyJSONPost,
-					p => p?.title.eq("whatever, it doesn't matter.")
+				syncExpression(LegacyJSONPost, p =>
+					p?.title.eq("whatever, it doesn't matter.")
```
Comment (stocaaro) [inline packages/datastore/__tests__/connectivityHandling.test.ts:2190]: Auto-formatting change

File: packages/datastore/src/sync/outbox.ts [hunk 2]
```diff
@@ -193,10 +193,12 @@ class MutationEventOutbox {
 		// Don't sync the version when the data in the response does not match the data
 		// in the request, i.e., when there's a handled conflict
 		//
-		// NOTE: `incomingData` contains all the fields in the record, and `outgoingData`
-		// only contains updated fields, resulting in an error when doing a comparison
-		// of two equal mutations. Fix this, or mitigate otherwise.
```
Comment (stocaaro) [inline packages/datastore/src/sync/outbox.ts:198]: This comment essentially describes what was needed here. Changing it to describe how/why this update works

File: packages/datastore/src/util.ts [hunk 3]
```diff
@@ -473,6 +473,32 @@ export function sortCompareFunction<T extends PersistentModel>(
...
+	valB: object,
+	nullish: boolean = false
+) {
+	const aKeys = Object.keys(valA);
+
+	for (const key of aKeys) {
+		const aVal = valA[key];
+		const bVal = valB[key];
+
+		if (!valuesEqual(aVal, bVal, nullish)) {
```
Comment (stocaaro) [inline packages/datastore/src/util.ts:494]: Doing a deep bi-directional equality check on all fields keeps parity with what existed before and works in all cases I've experimented with. I have a couple cases I intend to get under integ testing today. Happy to have suggestions if others have thoughts on cases to cover.

**Review by svidgen — CHANGES_REQUESTED** [review]
There's at least one case that still doesn't make sense to me.

https://github.com/aws-amplify/amplify-js/blob/ebf57ef182747f5ceff89308a3d05da1aff30122/packages/datastore/__tests__/connectivityHandling.test.ts#L1570-L1582

If everything is working correctly, I wouldn't expect `post title 0` to resurface at the end there.

**Comment by stocaaro** [comment]
The case testcase pointed out by Jon is interesting. I had interpreted it incorrectly as being caused because the fake service wasn't feeding updates from the second client back to the DS client. The Fake service is much more accurate than I was thinking/expecting.

Whats actually happening here is that the version number in the fake service is incremented the version number on failures. The second client is posting an update with an old version, which isn't accepted, but does increment the version number. This incremented version number doesn't make it out to the DS client before the subsequent updates are attempted, so they also fail, (again incrementing the version number), which is how we end up with a final version number of 5.

From the service perspective:
- `['original title', 1]` - Succeed in creating post with title 'original title'
- `['post title 0', 2]` - Succeed in updating post with title 'post title 0'
- `['post title 0', 3]` - Fail to update post with title 'update from second client'
- `['post title 0', 4]` - Fail to update post with title 'post title 1'
- `['post title 0', 5]` - Fail to update post with title 'post title 2'

**Pull request closed without merging**